src/createStore.ts

```diff
@@ -8,7 +12,36 @@ let createStoreAction = action('createStore', function createStoreAction(
     getRootStore().set(key, initialState);
 });
 
-export default function createStore<T>(key: string, initialState: T): () => T {
+export default function createStore<TState>(
+    key: string,
+    arg2: TState | LeafMutator<TState> | CombinedMutator<TState>
+): () => TState {
+    // Get the initial state (from the mutator, if necessary)
+    let mutator = getMutator(arg2);
+    let initialState = mutator ? mutator.getInitialValue() : arg2;
+
+    // Create the store under the root store
     createStoreAction(key, initialState);
-    return () => <T>getRootStore().get(key);
+    let getStore = () => <TState>getRootStore().get(key);
+
+    // If necessary, hook the mutator up to the dispatcher
+    if (mutator) {
+        subscribeAll(
+            wrapMutator(actionMessage => {
+                mutator.handleAction(getStore(), actionMessage, newState => {
+                    getRootStore().set(key, newState);
+                });
+            })
+        );
+    }
+
+    return getStore;
+}
+
+function getMutator<TState>(mutator: TState | LeafMutator<TState> | CombinedMutator<TState>) {
+    if ((<any>mutator).handleAction) {
+        return <LeafMutator<TState> | CombinedMutator<TState>>mutator;
```

Comment:
Would it make sense for `LeafMutator` and `CombinedMutator` to implement a common interface?

Reply:
I went back and forth on this...  The problem is that `LeafMutator` has an extra method that `CombinedMutator` doesn't, so I still need the package to export that type.  I could create an interface for the stuff that's common, but it's just one more type that needs to be exported.

---
In reply to: [200791352](https://github.com/Microsoft/satcheljs/pull/100#discussion_r200791352) [](ancestors = 200791352)